pkg/urbit/vere/pier.c

```diff
@@ -652,167 +596,136 @@ _pier_apply(u3_pier* pir_u)
+    u3x_trel(dat, &who, &fak, &len);
 
-  if ( mug_l != u3r_mug(wit_u->mat) ) {
-    //  fprintf(stderr, "pier: load: precommit: reveal failed: %s\r\n", nam_c);
-    u3_foil_delete(0, 0, wit_u->fol_u);
-    u3z(wit_u->mat);
-    c3_free(wit_u);
+    c3_assert( c3y == u3ud(who) );
+    c3_assert( 1 >= u3r_met(7, who) );
+    c3_assert( c3y == u3ud(fak) );
+    c3_assert( 1 >= u3r_met(0, fak) );
+    c3_assert( c3y == u3ud(len) );
+    c3_assert( 1 >= u3r_met(3, len) );
 
-    return 0;
-  }
+    _pier_set_ship(pir_u, u3k(who), u3k(fak));
 
-  /*  the problem with a precommit is that we don't know whether we
-  **  actually computed and acknowledged it.  the worst case is a
-  **  network packet that causes an infinite loop, which we crash
-  **  while trying to execute.
-  **
-  **  if we discard a precommit that in fact completed, we may have
-  **  sent an acknowledgment without realizing it -- creating a
-  **  network discontinuity.  but if we uniformly apply all precommits,
-  **  we may "apply" an infinite loop.  there is no perfect answer,
-  **  of course.
-  **
-  **  the correct behavior here is a consistent mapping from event
-  **  to timer.  we apply this same mapping, but quadruple the timer
-  **  expiration, while re-executing precommits.  if it still times
-  **  out, we conclude that we must not have completed the original
-  **  event, and can throw away the precommit.
-  */
+    pir_u->lif_d = u3r_chub(0, len);
+  }
 
-  return wit_u;
+  u3z(dat);
 }
 
-/* _pier_compare(): ascending sort compare.
+/* _pier_disk_read_header():
+** XX async
+** XX very slow
 */
-static c3_i
-_pier_compare(const void* vod_p, const void* dov_p)
+static void
+_pier_disk_read_header(u3_disk* log_u)
 {
-  const u3_writ* const* wit_u = vod_p;
-  const u3_writ* const* twi_u = dov_p;
+  c3_assert( 0 != log_u->fol_u );
 
-  return ((c3_ds)((*wit_u)->evt_d) - (c3_ds)((*twi_u)->evt_d));
-}
+  c3_d pos_d = log_u->fol_u->end_d;
+  c3_o got_o = c3n;
 
-/* _pier_disk_load_precommit(): load all precommits.
-*/
-static u3_writ**
-_pier_disk_load_precommit(u3_pier* pir_u,
-                          c3_d     lav_d)
-{
-  u3_disk* log_u = pir_u->log_u;
-  u3_dent* all_u = log_u->pre_u->all_u;
-  u3_writ* pre_u = 0;
-  c3_w     num_w = 0;
-
-  while ( all_u ) {
-    u3_writ* wit_u = _pier_disk_load_precommit_file(pir_u,
-                                                    lav_d,
-                                                    all_u->nam_c);
-
-    if ( wit_u ) {
-      wit_u->nex_u = pre_u;
-      pre_u = wit_u;
-      num_w++;
+  c3_assert( 0ULL != pos_d );
+
+  while ( pos_d ) {
+    c3_d  len_d, evt_d;
+    c3_d* buf_d;
+    u3_noun mat, ovo, job, evt;
+
+    buf_d = u3_foil_reveal(log_u->fol_u, &pos_d, &len_d);
+
+    if ( !buf_d ) {
+      _pier_disk_bail(0, "corrupt header");
+      return;
     }
-    all_u = all_u->nex_u;
-  }
 
-  {
-    u3_writ** ray_u = c3_malloc((1 + num_w) * sizeof(u3_writ*));
-    c3_w      i_w;
+    if ( 0ULL == pos_d) {
```

Comment:
I feel like this would be clearer if moved to after the while loop, unless I'm misunderstanding it.

Reply:
The way this function is written, it needs an early return. There used to be an event in this position. I agree this should be rewritten, if we are going to keep it. But I'm not sure if we're going to keep it.